### PR TITLE
Lobbies

### PR DESCRIPTION
**Pull request by Indexu**

### CODE REVIEW

**Comment by Indexu** [comment]
@judgeaxl You can merge this if you don't find anything odd about it.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: Could use MoveTemp here to avoid a copy

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: The CurrentLobby URLs were supposed to be removed, since they require potentially costly lookup on the root resource, which is supposed to be info only. They'll need to be filled in after an initial query, if the game needs it.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: I suggest replacing "weird" here with a description of the action that will be taken, i.e. resetting the lobby (session) or something like that.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: If this is the first thing that happens, will anyone know about this lobby and care about this notification? Might just be confusing if I start up a new game and get told my lobby is missing. Depends on who listens to this and what they do with it of course.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: I can't help but feeling that the name `ExtractLobby` suggests it would return the lobby information, not set it internally. `SetLobbyDataFromResponse` perhaps?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:97]: Should the CurrentLobbyURL members not be reset here?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: There's another configuration mode called "test" which is an optimized build with development features, and which also supports the console, so you might want to change the expression to not shipping. Not required, just a FYI

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:230]: There's been a debate regarding if it makes sense to fire this delegate or not, when the task never gets started, as you have to deal with a different ordering as a caller. The delegate will normally fire success or failure after the initial call has returned, meaning the caller has been able to store state, configure handlers, and what else. When preconditions fail, the callback happens before the initial call has returned, so the handler needs to be careful in what context it processes the result.

In new code I suggest we rely on the bool return to indicate "failed to initiate the request", and the delegate only to be used to signal the result of the request if it actually got started.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:280]: Should the lobby state be reset here?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: The pattern to use, if we have to manually construct URLs, and we should still avoid that as much as possible, is to expose a template URL in the root resource, like `lobby_members_template = https://foo.fg-api.com/drift/lobbies/{lobby_id}/members` and use replace on the URL before using it, and example can be found in `FDriftBase::SavePlayerGameStateImpl`.

In this particular case, where do we get the Lobby ID from? Can't it return the relevant URL too?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:350]: If CurrentLobbyId is empty here, but CurrentLobbyMemberURL is not, what's the state? CurrentLobbyId is used here and below even if it's determined to be empty.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: If the request later fails below, won't this be out of sync?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: Missing a return/fail here?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: Until replaced with a template, per the above use case, this should be separated out to a method to avoid duplication

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:676]: The delegate should be fired here too for consistency with the failure cases, but on the other hand, as per not at the top of the file, perhaps the delegates should not fire at all if the request never fires.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:710]: Is it possible to recover from this error or is the lobby now failed forever?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:735]: Are all lobby events supposed to contain a lobby_id? If so, I don't think this should query the lobby.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:743]: I'm not entirely happy with all the QueryLobby calls when the data is not matching expectations. Is there a reason why these events would contain the wrong data?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:806]: Would it not make sense to pass along some information about the player that joined?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:819]: Would it not make sense to pass along some information about the player that got updated?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:832]: Would it not make sense to pass along some information about the player that got deleted?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp:845]: Would it not make sense to pass along some information about the player that got kicked?

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: When using non-braced single line ifs like this, I suggest instead writing:

`if (condition) { return result; }`

as that is resistant to future refactorings which might introduce additional lines, plus it's even more compact.

File: Drift/Private/DriftLobbyManager.cpp
Comment (judgeaxl) [inline Drift/Private/DriftLobbyManager.cpp]: Same problem as ExtractLobby, the name suggests that it would return the members, not magically set them internally

File: Drift/Private/DriftSchemas.cpp
Comment (judgeaxl) [inline Drift/Private/DriftSchemas.cpp]: As per separate discussion, these are problematic

**Review by judgeaxl — CHANGES_REQUESTED** [review]

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:280]: Probably

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: We get the ID from user input, so impossible to get the URL beforehand via some drift-base response.

Will take a look at the SavePlayerGameStateImpl

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:350]: With the removal of my_xyz URLs, the state is consistent and I don't think it's possible to get in such a state.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Will rework this.
Same with update lobby.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Perhaps.
This kick code has never been run before since I didn't have time to add it to the UI and properly test it prior to the tournament.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:676]: Then should the return value be false? Since the request never fires.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:710]: You can recover from this, yes.
It might not be clear, but this happens when the start match failed. You can retry starting the match.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:735]: They are all supposed to, yes.
In the case that for some mystical reason, drift-base doesn't provide a message in the lobby message queue with a lobby id, then the message is borked, this logs out and syncs with the server the current lobby state via QueryLobby.

Then we get notified of this via Sentry and should take a look at what happened.
Ideally, this should never be run.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:743]: Paranoia.

Also, if someone in the future makes changes to the backend and doesn't update the C++ to follow along, we get a bunch of logs saying stuff is borked.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:806]: It would, yes.
We can add that in the future if we need it.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:819]: It would, yes.
We can add that in the future if we need it.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:832]: It would, yes.
We can add that in the future if we need it.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:845]: It would, yes.
We can add that in the future if we need it.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: It was a copy pasta job from somewhere else in Drift.

I can add the braces if you want.

File: Drift/Private/DriftSchemas.cpp
Comment (Indexu) [reply Drift/Private/DriftSchemas.cpp]: Removed

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Removed

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Remnant from copy pasting the query lobby code.
Removed.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: I don't know where I got that habit, but for some reason I always name functions that parse and set member variables from some response/arg/input/output `ExtractXYZ`.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:97]: After removing the my urls, this is no longer an issue.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp:230]: hmm...
I'm using the third param in the delegate as the error string which is then used in UI to notify the user what went wrong. Could also be a status code and looked up in some localization file, but same issue.
Hard to know what goes wrong just from 'false'.

Ideas?

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Added template urls to drift-base and string replace to construct the URLs.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Renamed to CacheLobby and CacheMembers.

Saw that terminology was used elsewhere in Drift.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Renamed to CacheLobby and CacheMembers.

Saw that terminology was used elsewhere in Drift.

File: Drift/Private/DriftLobbyManager.cpp
Comment (Indexu) [reply Drift/Private/DriftLobbyManager.cpp]: Added the braces. Looks nifty.

**Review by judgeaxl — APPROVED** [review]

**Pull request merged**
